File: csharp/BoundedLoad.cs
namespace ConsistentHash;

/// <summary>
/// Consistent hashing with bounded loads (Google, 2017): keys are placed clockwise on
/// the ring, but no node may exceed ceil((1+epsilon)*keys/nodes). A full node overflows
/// to the next under cap, so no node exceeds its fair share by more than epsilon.
/// Placement depends on load, so it is stateful and order-sensitive.
/// </summary>
public sealed class BoundedLoad
{
    private readonly double _epsilon;
    private readonly List<string> _nodes;
    private readonly SortedDictionary<uint, string> _ring = new();
    private uint[] _sorted = Array.Empty<uint>();
    private readonly Dictionary<string, int> _load = new();
    private readonly Dictionary<string, string> _assigned = new();

    public BoundedLoad(IEnumerable<string>? nodes = null, int replicas = 100, double epsilon = 0.25)
    {
        if (epsilon < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(epsilon), "epsilon must be non-negative");
        }

        _epsilon = epsilon;
        _nodes = (nodes ?? Enumerable.Empty<string>())
            .Distinct()
            .OrderBy(n => n, StringComparer.Ordinal)
            .ToList();
        BuildRing(replicas);
    }

    public string? Get(string key)
    {
        if (_sorted.Length == 0)
        {
            return null;
        }

        if (_assigned.TryGetValue(key, out var existing))
        {
            return existing;
        }

        var cap = Capacity();
        var start = FirstClockwise(HashRing.Fnv1a32(key));
        for (var i = 0; i < _sorted.Length; i++)
        {
            var node = _ring[_sorted[(start + i) % _sorted.Length]];
            if (_load.GetValueOrDefault(node, 0) < cap)
            {
                Assign(key, node);
                return node;
            }
        }

        var fallback = _ring[_sorted[start % _sorted.Length]];
        Assign(key, fallback);
        return fallback;
    }

    public IReadOnlyDictionary<string, int> Loads => new Dictionary<string, int>(_load);

    public IReadOnlyList<string> Nodes => _nodes.ToList();

    public int Count => _nodes.Count;

    private void Assign(string key, string node)
    {
        _assigned[key] = node;
        _load[node] = _load.GetValueOrDefault(node, 0) + 1;
    }

    private int Capacity()
    {
        var total = _assigned.Count + 1;
        return (int)Math.Ceiling((1 + _epsilon) * total / _nodes.Count);
    }

    private void BuildRing(int replicas)
    {
        foreach (var node in _nodes)
        {
            for (var r = 0; r < replicas; r++)
            {
                var slot = HashRing.Fnv1a32($"{node}#{r}");
                while (_ring.ContainsKey(slot))
                {
                    slot += 1;
                }

                _ring[slot] = node;
            }
        }

        _sorted = _ring.Keys.ToArray();
    }

    private int FirstClockwise(uint h)
    {
        int lo = 0, hi = _sorted.Length;
        while (lo < hi)
        {
            var mid = (lo + hi) / 2;
            if (_sorted[mid] <= h)
            {
                lo = mid + 1;
            }
            else
            {
                hi = mid;
            }
        }

        return lo == _sorted.Length ? 0 : lo;
    }
}

File: java/src/main/java/com/consistenthash/BoundedLoad.java
package com.consistenthash;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.TreeSet;

/**
 * Consistent hashing with bounded loads (Google, 2017): keys are placed clockwise on
 * the ring, but no node may exceed ceil((1+epsilon)*keys/nodes). A full node overflows
 * to the next under cap, so no node exceeds its fair share by more than epsilon.
 * Placement depends on load, so it is stateful and order-sensitive.
 */
public final class BoundedLoad {

    private final double epsilon;
    private final List<String> nodes;
    private final TreeMap<Long, String> ring = new TreeMap<>();
    private final Map<String, Integer> load = new HashMap<>();
    private final Map<String, String> assigned = new HashMap<>();

    public BoundedLoad() {
        this(new ArrayList<>(), 100, 0.25);
    }

    public BoundedLoad(List<String> nodes, int replicas, double epsilon) {
        if (epsilon < 0) {
            throw new IllegalArgumentException("epsilon must be non-negative");
        }
        this.epsilon = epsilon;
        this.nodes = new ArrayList<>(new TreeSet<>(nodes));
        buildRing(replicas);
    }

    public String get(String key) {
        if (ring.isEmpty()) {
            return null;
        }
        String existing = assigned.get(key);
        if (existing != null) {
            return existing;
        }
        int cap = capacity();
        long h = HashRing.fnv1a32(key);
        // Walk clockwise from the key, wrapping once, taking the first node under cap.
        List<String> order = new ArrayList<>(ring.tailMap(h, false).values());
        order.addAll(ring.values());
        for (String node : order) {
            if (load.getOrDefault(node, 0) < cap) {
                assign(key, node);
                return node;
            }
        }
        String fallback = order.get(0);
        assign(key, fallback);
        return fallback;
    }

    public Map<String, Integer> loads() {
        return new HashMap<>(load);
    }

    public List<String> nodes() {
        return new ArrayList<>(nodes);
    }

    public int size() {
        return nodes.size();
    }

    private void assign(String key, String node) {
        assigned.put(key, node);
        load.merge(node, 1, Integer::sum);
    }

    private int capacity() {
        int total = assigned.size() + 1;
        return (int) Math.ceil((1 + epsilon) * total / nodes.size());
    }

    private void buildRing(int replicas) {
        for (String node : nodes) {
            for (int r = 0; r < replicas; r++) {
                long slot = HashRing.fnv1a32(node + "#" + r);
                while (ring.containsKey(slot)) {
                    slot = (slot + 1) & 0xFFFFFFFFL;
                }
                ring.put(slot, node);
            }
        }
    }
}
